core/src/main/scala/org/apache/spark/scheduler/Pool.scala

```diff
@@ -119,4 +119,28 @@ private[spark] class Pool(
       parent.decreaseRunningTasks(taskNum)
     }
   }
+
+  override def updateAvailableSlots(numSlots: Float): Unit = {
+    schedulingMode match {
+      case SchedulingMode.FAIR =>
+        val usableWeights = schedulableQueue.asScala
+          .map(s => if (s.getSortedTaskSetQueue.nonEmpty) (s, s.weight) else (s, 0))
+        val totalWeights = usableWeights.map(_._2).sum
+        usableWeights.foreach({case (schedulable, usableWeight) =>
```

Comment:
also I don't think this is correct.  The numSlots is the number of currently available slots (ignoring used), here you try to compute what each one gets of those slots based on minshare/weight, but you aren't taking into account what each pool is already using.  I might have one pool already using all the slots it should be and one pool using none.  The one using all of its slots should get 0 available for it. Although that probably isn't quite right either because the weights and minshares just control the order of the tasksets so I think it can actually get more depending on other factors.  Like taskset that is highest prioirty hasn't met locality wait, so then goes to the taskset that is using more but has met locality wait or doesn't have.  Then it will could grab one which would at least temporarily put it over its fair share.  
This is why Kay was saying its much harder here and was talking about using proxy's to help indicate the conditions.

Reply:
this is computing how many slots each taskset/pool could ideally use if distributed according to scheduling policy, not how many **more slots** it can use in the next round of scheduling. 

Say there are 10 total slots with two jobs running, each which have the same weight, so they are assigned a potential of 5 slots each. Also assume taskset 1 is using 6 slots and taskset 2 is using 0.
This code will determine that each taskset's availableSlots is 5, and since taskset 1 is using more than that, its timer will reset, whereas taskset 2's will not. This means taskset 2 will increase its locality level if the locality timer has expired, allowing it to utilize its "ideal" 5 slots.

Reply:
Oh I see, I misread when  executorIdToCores(o.executorId) = o.cores was set, I thought it was set on every iteration and updated with free cores not the total.